### DISCUSSION TITLE
Suitability for automated code refactoring (codemods)

### DISCUSSION THREAD

**Discussion by danielkza** [tree-sitter/tree-sitter#1108] [Q&A] [answered]
Hi folks,

I am looking to learn about different open-source tools available to perform automated source code refactoring. While from reading the documentation tree-sitter seems not have been created with such goal, it's ability to manipulate concrete syntax trees seems like a very strong building block to start with.

Has anyone used it in such fashion? How suitable would the available APis be for building some tooling with that goal?

Thanks in advance, and apologies if a similar question has already been answered, but I could not find anything about it while Googling.

**Comment by maxbrunsfeld** [comment, marked as answer]
Tree-sitter should be suitable for these kinds of tools because as you said it gives you a full concrete syntax tree (as opposed to an AST), but it does try to present the CST in a way that analysis is still easy, like it would be with an AST, using features like [fields](https://tree-sitter.github.io/tree-sitter/using-parsers#node-field-names) and [queries](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries).

One limitation that you should be aware of is that Tree-sitter doesn't have APIs for *mutating* syntax trees, and then reflecting those changes back as text diffs. That's something  you'd need to build yourself.

**Reply by marcel0ll** [reply]
@maxbrunsfeld what are the reasons that tree-sitter does not have mutating apis? Would it be possible to add it?

**Reply by ahlinc** [reply]
@marcel0ll you may find more context with answers in the [several](https://github.com/tree-sitter/tree-sitter/issues?q=mutate+OR+mutating+OR+mutation+OR+mutability+OR+mutable+in%3Atitle+in%3Atitle%2Cbody) [issues](https://github.com/tree-sitter/tree-sitter/issues?q=mutate+OR+mutating+OR+mutation+OR+mutability+OR+mutable+in%3Atitle+in%3Atitle%2Cbody%2Ccomments) and [discussions](https://github.com/tree-sitter/tree-sitter/discussions?discussions_q=mutate+OR+mutating+OR+mutation+OR+mutability+OR+mutable+in%3Atitle%2Cbody%2Ccomments).
